File: backend/src/main/java/com/lab1/lab1/model/model/Movie.java
package com.lab1.lab1.model.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import lombok.Data;

@Entity
@Data
public class Movie {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id; //Значение поля должно быть больше 0, Значение этого поля должно быть уникальным, Значение этого поля должно генерироваться автоматически

    @Column(nullable = false)
    @NotBlank
    private String name; //Поле не может быть null, Строка не может быть пустой

    @ManyToOne
    private Coordinates coordinates; //Поле не может быть null

    @Column(nullable = false)
    @Temporal(TemporalType.TIMESTAMP)
    private java.util.Date creationDate; //Поле не может быть null, Значение этого поля должно генерироваться автоматически

    @Column(nullable = false)
    @Min(1)
    private int oscarsCount; //Значение поля должно быть больше 0

    @Column(nullable = false)
    @Min(0)
    private float budget; //Значение поля должно быть больше 0

    @Column(nullable = false)
    @Min(0)
    private Double totalBoxOffice; //Поле не может быть null, Значение поля должно быть больше 0

    @Enumerated(EnumType.STRING)
    private MpaaRating mpaaRating; //Поле может быть null

    @ManyToOne(optional = false)
    private Person director; //Поле не может быть null

    @ManyToOne
    private Person screenwriter;

    @ManyToOne(optional = false)
    private Person operator; //Поле не может быть null

    @Column(nullable = false)
    @Min(1)
    private long length; //Значение поля должно быть больше 0

    @Column(nullable = false)
    @Min(1)
    private int goldenPalmCount; //Значение поля должно быть больше 0

    @Min(0)
    private Long usaBoxOffice; //Поле может быть null, Значение поля должно быть больше 0

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private MovieGenre genre; //Поле не может быть null

}

File: frontend/lib/features/main/domain/entities/movie.dart
import 'person.dart';

class Movie {
  int id;
  String name;
  Coordinates coordinates;
  DateTime creationDate;
  int oscarsCount;
  double budget;
  double totalBoxOffice;
  MpaaRating mpaaRating;
  Person director;
  Person screenwriter;
  Person operator;
  int length;
  int goldenPalmCount;
  double usaBoxOffice;
  MovieGenre genre;

  Movie({
    required this.id,
    required this.name,
    required this.coordinates,
    required this.creationDate,
    required this.oscarsCount,
    required this.budget,
    required this.totalBoxOffice,
    required this.mpaaRating,
    required this.director,
    required this.screenwriter,
    required this.operator,
    required this.length,
    required this.goldenPalmCount,
    required this.usaBoxOffice,
    required this.genre,
  });

  String genreToString() => switch (genre) {
        MovieGenre.ACTION => 'Экшен',
        MovieGenre.FANTASY => 'Фэнтези',
        MovieGenre.WESTERN => 'Вестерн'
      };

  String mpaaRatingToString() => mpaaRating.toString().split('.').last;
}

class Coordinates {
  double x;
  double y; //Поле не может быть null

  Coordinates({
    required this.x,
    required this.y,
  });
}

enum MpaaRating {
  G,
  PG_13,
  NC_17;
}

enum MovieGenre {
  ACTION,
  WESTERN,
  FANTASY;
}
